sdk/metric/aggregator/array/array.go

```diff
@@ -0,0 +1,99 @@
+// Copyright 2019, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package array
+
+import (
+	"context"
+	"sort"
+	"sync"
+
+	"go.opentelemetry.io/api/core"
+	"go.opentelemetry.io/sdk/export"
+)
+
+type (
+	Aggregator struct {
+		lock   sync.Mutex
+		points Points
+		saved  Points
+		sum    core.Number
+	}
+
+	Points []core.Number
+)
+
+var _ export.MetricAggregator = &Aggregator{}
+
+func New() *Aggregator {
+	return &Aggregator{}
+}
+
+func (a *Aggregator) Collect(ctx context.Context, rec export.MetricRecord, exp export.MetricBatcher) {
+	a.lock.Lock()
+	a.saved, a.points = a.points, nil
+	a.lock.Unlock()
```

Comment:
How do all the layers fit together here?  Does anything "own" or have sole access to an aggregator?

Reply:
The SDK has a lock-free algorithm that (I assert) maintains records until they are no longer needed.  Inside a call to SDK.Collect() it will call `Collect()` on each aggregator.  If the aggregator was stale, it'll be removed from memory after reporting a final time.  If there's a race, we end up with multiple records for the same metric.